File: Iteration_2/src/main/java/Transcript.java
import java.util.ArrayList;

public class Transcript {
    private ArrayList<Course> completedCourses;
    private ArrayList<Course> currentCourses;
    private ArrayList<Course> waitedCourses;
    private WeeklySchedule weeklySchedule;
    private ArrayList<CourseSection> currentSections;
    private ArrayList<CourseSection> waitedSections;
    private int semester;

    public Transcript(){

    };

    public Transcript(ArrayList<Course> comp, ArrayList<Course> current){
        this.completedCourses = comp;
        this.currentCourses = current;
    };
//constructor for Transcript object
    public Transcript(ArrayList<Course> completedCourses, ArrayList<Course> currentCourses, ArrayList<Course> waitedCourses){
        this.completedCourses = completedCourses;
        this.currentCourses = currentCourses;
        this.waitedCourses = waitedCourses;
        this.currentSections = addCourseSections(currentCourses);
        this.waitedSections = addCourseSections(waitedCourses);
       
    }
    public Transcript(ArrayList<Course> completedCourses, ArrayList<Course> currentCourses, ArrayList<Course> waitedCourses, ArrayList<CourseSection> currentSections, ArrayList<CourseSection> waitedSections) {
        this.completedCourses = completedCourses;
        this.currentCourses = currentCourses;
        this.waitedCourses = waitedCourses;
        this.currentSections = currentSections;
        this.waitedSections = waitedSections;
    }
//This method adds a course to the list of completed courses if it has been completed.
    public void addCompletedCourse(Course course){
        try {
            if (course != null)
                completedCourses.add(course);
            else
                System.out.println("The course object is null");
        } catch (NullPointerException e) {
            System.out.println("The completedCourses list has not been initialized.");
        }catch (Exception e1){
            System.out.println(e1.getMessage());
        }
    }
//This method adds the courses approved by the advisor from the waiting list to the currently taken courses list.
    public void addCurrentCourse(Course course){
        try {
            if (course != null)
                currentCourses.add(course);
            else
                System.out.println("The course object is null");
        } catch (NullPointerException e) {
            System.out.println("The completedCourses list has not been initialized.");
        }catch (Exception e1){
            System.out.println(e1.getMessage());
        }
    }
    //Adds the courses selected by the student and sent for approval to the waiting course list
    public void addWaitedCourse(Course course){
        try {
            if (course != null)
                waitedCourses.add(course);
            else
                System.out.println("The course object is null");
        } catch (NullPointerException e) {
            System.out.println("The waitedCourses list has not been initialized.");
        }catch (Exception e1){
            System.out.println(e1.getMessage());
        }
    }
//This method deletes the courses from the waiting course list according to the approval or rejection status of the place where it is called.
    public void deleteFromWaitedCourse(Course course){
        try {
            if (course != null)
                waitedCourses.remove(course);
            else
                System.out.println("The course object is null");
        } catch (NullPointerException e) {
            System.out.println("The waitedCourses list has not been initialized.");
        }catch (Exception e1){
            System.out.println(e1.getMessage());
        }
    }
    public void addCurrentSection(CourseSection courseSection){
        try {
            if (courseSection != null)
                currentSections.add(courseSection);
            else
                System.out.println("The course object is null");
        } catch (NullPointerException e) {
            System.out.println("The waitedCourses list has not been initialized.");
        }catch (Exception e1){
            System.out.println(e1.getMessage());
        }
    }

    public void deleteFromWaitedSections(CourseSection courseSection){
        try {
            if (courseSection != null)
                waitedSections.remove(courseSection);
            else
                System.out.println("The sections object is null");
        } catch (NullPointerException e) {
            System.out.println("The waitedSections list has not been initialized.");
        }catch (Exception e1){
            System.out.println(e1.getMessage());
        }
    }

//Scans the waitedCourses list and prints the courses there
    public void showWaitedCourses(){
        System.out.println("\nCourses that are waiting for approve listed below:");
        for(int i = 0; i < waitedCourses.size(); i++){
            System.out.printf("%-3d-   %-10s %-50s %d\n", (i+1), waitedCourses.get(i).getCourseId(), waitedCourses.get(i).getCourseName(), waitedCourses.get(i).getCredits());
        }
    }
//Scans the completedCourses list and prints the courses there
    public void showCompletedCourses(){
        System.out.println("\nCourses that are completed listed below:");
        for(int i = 0; i < completedCourses.size(); i++){
            System.out.printf("%-3d-   %-10s %-50s %-10d %s\n", (i+1), completedCourses.get(i).getCourseId(), completedCourses.get(i).getCourseName(), completedCourses.get(i).getCredits(), completedCourses.get(i).getGrade());
        }
    }
//Scans the currentCourses list and prints the courses there
    public void showCurrentCourses(){
        System.out.println("\nCourses that are registered listed below:");
        for(int i = 0; i < currentCourses.size(); i++){
            System.out.printf("%-3d-   %-10s %-50s %d\n", (i+1), currentCourses.get(i).getCourseId(), currentCourses.get(i).getCourseName(), currentCourses.get(i).getCredits());
        }
    }
   public ArrayList<CourseSection> addCourseSections(ArrayList<Course> courses){
        int size = courses.size();
        ArrayList<CourseSection> cs = new ArrayList<CourseSection>();
        for(int k=0; k < size; k++ ){
            for(int j = 0; j < courses.get(k).getCourseSections().size(); j++){
                cs.add(courses.get(k).getCourseSections().get(j));
            }
            
        }
        return cs;
   }
    
//Sets the list of courses that the student has completed
    public void setCompletedCourses(ArrayList<Course> completedCourses) {
        this.completedCourses = completedCourses;
    }
//Sets the list of courses that the student is currently enrolled in
    public void setCurrentCourses(ArrayList<Course> currentCourses) {
        this.currentCourses = currentCourses;
    }
//Sets the list of courses that the student is waiting to enroll in
    public void setWaitedCourses(ArrayList<Course> waitedCourses) {
        this.waitedCourses = waitedCourses;
    }
//Retrieves the list of courses that the student has completed.
    public ArrayList<Course> getCompletedCourses() {
        return completedCourses;
    }
//Retrieves the list of courses that the student is currently enrolled in
    public ArrayList<Course> getCurrentCourses() {
        return currentCourses;
    }

    // Retrieves the list of courses that the student is waiting to enroll in.
    // If the waitlist is null, returns an empty ArrayList   

    public ArrayList<Course> getWaitedCourses() {
        return waitedCourses != null ? waitedCourses : new ArrayList<>();
    }

    public ArrayList<CourseSection> getCurrentSections() {
        return currentSections;
    }

    public void setCurrentSections(ArrayList<CourseSection> currentSections) {
        this.currentSections = currentSections;
    }

    public ArrayList<CourseSection> getWaitedSections() {
        return waitedSections;
    }

    public void setWaitedSections(ArrayList<CourseSection> waitedSections) {
        this.waitedSections = waitedSections;
    }

    public double calculateGpa(){
        int size = completedCourses.size();
        double totalSum = 0;
        int totalCredits = 0;

        for(int i = 0; i < size; i++){
            String grade = completedCourses.get(i).getGrade();
            int credits = completedCourses.get(i).getCredits();
            switch (grade) {
                case "AA": totalSum += 4.0 * credits;
                break;
                case "BA": totalSum += 3.5 * credits;
                break;
                case "BB": totalSum += 3.0 * credits;
                break;
                case "CB": totalSum += 2.5 * credits;
                break;
                case "CC": totalSum += 2.0 * credits;
                break;
                case "DC": totalSum += 1.5 * credits;
                break;
                case "DD": totalSum += 1.0 * credits;
                break;
                case "FD": totalSum += 0.5 * credits;
                break;
                case "FF": totalSum += 0.0 * credits;
                break;
                default:
            }
            totalCredits += credits;
        }
        return (int)((totalSum/totalCredits) * 100) / 100.0;
    }

    public int getSemester(){
        return this.semester;
    }
    
    public void setSemester(int semester){
        this.semester = semester;
    }
}

File: Iteration_3/Transcript.py
class Transcript:
    def __init__(self, completed_courses=None, current_courses=None, waited_courses=None, 
                 current_sections=None, waited_sections=None, semester=0):
        self.__completed_courses = completed_courses or []
        self.__current_courses = current_courses or []
        self.__waited_courses = waited_courses or []
        self.__current_sections = current_sections or []
        self.__waited_sections = waited_sections or []
        self.semester = semester

    def add_completed_course(self, course):
        if course:
            self.__completed_courses.append(course)
        else:
            print("The course object is null")

    def add_current_course(self, course):
        if course:
            self.__current_courses.append(course)
        else:
            print("The course object is null")

    def add_waited_course(self, course):
        if course:
            self.__waited_courses.append(course)
        else:
            print("The course object is null")

    def delete_from_waited_course(self, course):
        if course in self.__waited_courses:
            self.__waited_courses.remove(course)
        else:
            print("The course object is null")

    def add_current_section(self, course_section):
        if course_section:
            self.__current_sections.append(course_section)
        else:
            print("The course section object is null")

    def delete_from_waited_sections(self, course_section):
        if course_section in self.__waited_sections:
            self.__waited_sections.remove(course_section)
        else:
            print("The course section object is null")

    def show_waited_courses(self):
        print("\nCourses that are waiting for approval listed below:")
        for i, course in enumerate(self.__waited_courses, start=1):
            print(f"{i:3}-   {course.get_course_id():<10} {course.get_course_name():<50} {course.get_credits()}")

    def show_completed_courses(self):
        print("\nCourses that are completed listed below:")
        for i, course in enumerate(self.__completed_courses, start=1):
            print(f"{i:3}-   {course.get_course_id():<10} {course.get_course_name():<50} {course.get_credits():<10} {course.get_grade()}")

    def show_current_courses(self):
        print("\nCourses that are registered listed below:")
        for i, course in enumerate(self.__current_courses, start=1):
            print(f"{i:3}-   {course.get_course_id():<10} {course.get_course_name():<50} {course.get_credits()}")

    def add_course_sections(self, courses):
        course_sections = []
        for course in courses:
            course_sections.extend(course.get_course_sections())
        return course_sections

    def set_completed_courses(self, completed_courses):
        self.__completed_courses = completed_courses

    def set_current_courses(self, current_courses):
        self.__current_courses = current_courses

    def set_waited_courses(self, waited_courses):
        self.__waited_courses = waited_courses

    def get_completed_courses(self):
        return self.__completed_courses

    def get_current_courses(self):
        return self.__current_courses

    def get_waited_courses(self):
        return self.__waited_courses

    def get_current_sections(self):
        return self.__current_sections

    def set_current_sections(self, current_sections):
        self.__current_sections = current_sections

    def get_waited_sections(self):
        return self.__waited_sections

    def set_waited_sections(self, waited_sections):
        self.__waited_sections = waited_sections

    def calculate_gpa(self):
        total_sum = 0
        total_credits = 0

        for course in self.__completed_courses:
            grade = course.get_grade()
            credits = course.get_credits()
            grade_to_gpa = {
                "AA": 4.0, "BA": 3.5, "BB": 3.0, "CB": 2.5, "CC": 2.0,
                "DC": 1.5, "DD": 1.0, "FD": 0.5, "FF": 0.0
            }
            total_sum += grade_to_gpa.get(grade, 0) * credits
            total_credits += credits

        return round(total_sum / total_credits, 2) if total_credits > 0 else 0.0

    def get_semester(self):
        return self.semester

    def set_semester(self, semester):
        self.semester = semester
